### PR TITLE
Add new clean, convert, export commands

### LINKED ISSUE #20: Clean command
May be useful to have an `nbautoexport` clean command to remove exports that don't match the existing configuration and notebooks. This would clean up exports after changing the configuration, or renaming notebooks.

### PR DESCRIPTION
**Pull request by jayqi**
Added new commands to the CLI. The summary of the commands are as follows:

- `clean`: Removes files that don't match what `nbautoexport` would create based on current notebooks and current config. For example, the use case is removing stuff like `Untitled.py` to clean up after stuff from renamed or deleted notebooks. 
- ~`convert`: This runs conversion with the configuration provided by CLI arguments (no .nbautoexport file needed).~ 
- `export`: This runs conversion on a folder with a .nbautoexport configuration file or with override CLI options.
- `install`: The old main command of creating a .nbautoexport configuration file. 

Additional changes:
- Adding an `clean` configuration option to automatically run the cleaning as part of `post_save`. Defaults to `False`.
- Add `rst` as an export format
- Add `__main__.py` to support for calling with `python -m`. This requires setting `typer>=0.3.0`.

Additional refactoring:
- Split source code into multiple modules
- Refactored export functionality into new function `export.export_notebook` to better support code reuse for `convert` and `export` commands.
- Refactored all instances of `os.path` to use `pathlib`
- Refactored .nbautoexport config to use pydantic model, with strict validation on fields

How stuff works:
- **How `clean`, ~`convert`~, and `export` know what files are notebooks**: 
  - New function `utils.find_notebook` will attempt to load every file in given directory with `nbformat`. 
- **How `clean` knows what files to expect**:
  - New function `clean.notebook_exports_generator` holds logic for predicting exported files. 
  - New function `clean.get_extension` grabs `nbconvert` exporters to figure out appropriate file extension

Fixed some bugs:
- The `asciidoc`, `latex`, `markdown`, and `rst` formats create separate directories for image files that weren't getting moved into the subfolders. 
- `pdf` format resulted in a unicode decode error because the postprocessor was trying to read the pdf file back in as unicode text in order to remove the cell numbers
- `nbconvert` grabs CLI arguments from `sys.argv`, which leads to really weird errors if `nbconvert` gets initiated through something like `pytest` or `nbautoexport`. Previously this was worked around in the tests by monkeypatching `sys.argv`. Instead, we now use a context manager to temporarily clear `sys.argv` whenever `nbconvert` is initialized, which prevents errors no matter if it was `nbautoexport`, `pytest`, or `jupyter` that starts the outer process. 

Looking for feedback:
- ~Are the names of the commands clear?~
  - ~`convert` and `export` could be confusing.~ 
  - ~`install` is what the original function for setting up was named this whole time. Are there better names? `configure`? `watch`? I personally like `watch`.~ Opened issue #34  
- ~Right now we only support `organize_by` of `notebooks` and `extensions`. Wondering if it would make sense to support `None`, i.e., keep converted output in parent directory and not in subfolders.~ Opened #32  

TODO:
- [X] Update unit tests to work with refactor
- [X] Add new unit tests for new commands

Known issues, thinking will save these for the future:
- [X] ~Determining appropriate file extension to get expected export files is currently clunky for `script`. `nbconvert` actually determines the language from loading the `.ipynb` file and then dispatches to the specific language exporter. We may want to do the same thing.~ _ended up implementing_ 
- [x] ~I don't actually know how to get the exporters for languages besides Python (expecting R and Julia?). The [`nbconvert` library](https://github.com/jupyter/nbconvert/tree/master/nbconvert/exporters) only directly supports `python`.  We'll have to figure out how it knows about others. Currently this will return `.txt`.~ Ran this against an R notebook I had, and it looks up the `.r` from the notebook metadata under `file_extension`. 
- [X] ~The expected export files functionality does not support the subdirectories that `asciidoc`, `markdown`, and `rst` create for images, so `clean` will remove them. This may require figuring out naming conventions, and/or refactoring the `expected_exports` method to return globs instead of Paths.~ _ended up implementing_
- [X] ~Functionality that searches for notebooks (`clean`, `export`, `convert`) currently just glob for `*.ipynb` files. This may miss stuff, and we should potentially instead try checking any file with [`nbformat`](https://nbformat.readthedocs.io/en/latest/) instead.~ _ended up implementing_

Note: I tried to get the root of the CLI to point to `install` but it doesn't work, because other subcommands will conflict with arguments. 

Resolves #20, #21

### CODE REVIEW

File: nbautoexport/clean.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    expected Jupyter or nbautoexport files.
+
+    Args:
+        directory (Path): notebooks directory to find files to clean up
+
+    Returns:
+        List[Path]: list of files to clean up
+    """
+    notebooks = find_notebooks(directory)
+    expected_exports = [directory / export for export in get_expected_exports(notebooks, config)]
```
Comment (jayqi) [inline nbautoexport/clean.py:103]: This `directory` is an extra and will lead to subfolder duplication. Add a unit test for this.

File: nbautoexport/nbautoexport.py [hunk 2]
```diff
@@ -1,263 +1,217 @@
...
 
-def initialize_post_save_hook(c: Config):
-    # >>> nbautoexport initialize, version=[{version}] >>>
-    try:
-        from nbautoexport import post_save
+    files_to_clean = find_files_to_clean(directory, config)
 
-        if callable(c.FileContentsManager.post_save_hook):
-            old_post_save = c.FileContentsManager.post_save_hook
+    typer.echo("Identified following files to clean up:")
```
Comment (jayqi) [inline nbautoexport/nbautoexport.py:76]: If `len(files_to_clean) == 0`, we should shortcircuit and exit the run.

File: nbautoexport/nbautoexport.py [hunk 3]
```diff
@@ -1,263 +1,217 @@
...
-            logger.info(f"Updating nbautoexport post-save hook with version {__version__}...")
-            with config_path.open("w") as fp:
-                # Open as w replaces existing file. We're replacing entire config.
-                fp.write(block_regex.sub(post_save_hook_initialize_block, config))
-        else:
-            logger.debug("No changes made.")
-            return
+    config = NbAutoexportConfig(export_formats=export_formats, organize_by=organize_by)
+    if input.is_dir():
+        for notebook_path in input.glob("*.ipynb"):
```
Comment (pjbull) [inline nbautoexport/nbautoexport.py:143]: Use `find_notebooks`

File: nbautoexport/nbautoexport.py [hunk 4]
```diff
@@ -1,263 +1,217 @@
...
-    sentinel_path = directory / SAVE_PROGRESS_INDICATOR_FILE
-
-    if sentinel_path.exists() and (not overwrite):
-        raise FileExistsError(
-            f"""Detected existing autoexport configuration at {sentinel_path}. """
-            """If you wish to overwrite, use the --overwrite flag."""
-        )
+    if input.is_dir():
+        sentinel_path = input / SAVE_PROGRESS_INDICATOR_FILE
+        notebook_paths = input.glob("*.ipynb")
```
Comment (pjbull) [inline nbautoexport/nbautoexport.py:166]: Use `find_notebooks`

**Comment by jayqi** [comment]
@pjbull @ejm714 @r-b-g-b 

Okay, I think this should be ready to go. 
- Merged the `export` and `convert` commands. This is just named `export`. 
- Using miniconda in the CI now, so that we can install pandoc. Works great across all three OSes. 
- I had to remove `pdf` from the unit tests because unfortunately it requires `xelatex`, and I couldn't figure out a way to easily install it in the CI. Will open a new issue about figuring this out. Ultimately, testing `pdf` had helped me catch a bug last week, so I would like to have it tested.

**Pull request merged**